src/index/PatternIndex.cpp

```diff
@@ -0,0 +1,834 @@
+
+const std::string PatternIndex::PATTERNS_NAMESPACE_SUFFIX =
+    ".index.patterns.pred_namespace";
+const std::string PatternIndex::PATTERNS_SUBJECTS_SUFFIX = ".index.patterns";
+const std::string PatternIndex::PATTERNS_OBJECTS_SUFFIX =
+    ".index.patterns.objects";
+
+// TODO: stxxl::vector<array<Id, 3>> should be Index::TripleVec but that
+// creates a circular dependency
+using TripleVec = stxxl::vector<array<Id, 3>>;
+
+PatternIndex::PatternIndex()
+    : _maxNumPatterns(std::numeric_limits<PatternID>::max() - 2),
+      _num_bytes_per_predicate(0),
+      _initialized(false) {}
+
+// _____________________________________________________________________________
+void PatternIndex::generatePredicateLocalNamespace(
+    VocabularyData* vocabData, const std::string& filename_base) {
+  // Create a new namespace for predicates containing only predicates.
+  // This will be significantly smaller than the global namespace which
+  // also contains subjects and objects, and allows for shrinking
+  // the pattern trick data.
+
+  // This is not inside a templated method as the PSO metadata is based upon
+  // HashMaps which need to be treated differently
+  TripleVec::bufreader_type reader(*vocabData->idTriples);
+  if (reader.empty()) {
+    LOG(WARN) << "Triple vector was empty, no patterns created" << std::endl;
+    return;
+  }
+
+  Id currentPred = ID_NO_VALUE;
+
+  Id langPredLowerBound = vocabData->langPredLowerBound;
+  Id langPredUpperBound = vocabData->langPredUpperBound;
+
+  // Iterate all triples in POS (or PSO) sorting order. Add every distinct
+  // non language predicate to the predicateIds vector, therefore assigning
+  // a predicate namespace id to it via its position in the vector.
+  for (; !reader.empty(); ++reader) {
+    Id predicate = (*reader)[1];
+    if (predicate != currentPred) {
+      currentPred = predicate;
+      if (predicate < langPredLowerBound || predicate >= langPredUpperBound) {
+        // The predicate is not a language predicate, add it to the ids
+        _predicate_local_to_global_ids.push_back(predicate);
+      }
+    }
+  }
+
+  // Compute the global to local mapping from the local to global mapping
+  _predicate_global_to_local_ids.reserve(_predicate_local_to_global_ids.size());
+  for (size_t i = 0; i < _predicate_local_to_global_ids.size(); ++i) {
+    _predicate_global_to_local_ids.try_emplace(
+        _predicate_local_to_global_ids[i], i);
+  }
+
+  writeNamespaceFile(filename_base + PATTERNS_NAMESPACE_SUFFIX,
+                     _predicate_local_to_global_ids);
+}
+
+// _____________________________________________________________________________
+void PatternIndex::generatePredicateLocalNamespaceFromExistingIndex(
+    Id langPredLowerBound, Id langPredUpperBound, IndexMetaDataHmap& meta_data,
+    const std::string& filename_base) {
+  // This is not inside a templated method as the PSO metadata is based upon
+  // HashMaps which need to be treated differently
```

Comment:
Do you mean: We only use this with PSO and PSO is always MetaDataHmap and thus we need no template?
I think this can be left out, it automatically breaks the compilation if we change this.
The only important INFO is that `meta_data` has to be sorted by the P column.